### DISCUSSION TITLE
Show and Tell: Are the Starting Materials for Synthesizing Your Target Molecules Commercially Available?

### DISCUSSION THREAD

**Discussion by bertiewooster** [rdkit/rdkit#6079] [Show and tell]
Given target molecules and reactions to synthesize them, determine whether the starting materials are commercially available using PubChem’s API, and thus whether the target is synthetically accessible.

![Three reactions, each in a row. First column: Target molecule and whether it's accessible based on commercial availability of reactants. Subsequent columns: Each reactant and whether it's commercial available.](https://user-images.githubusercontent.com/1335522/218208990-40a9d329-70df-416f-af61-e15532c0314b.png)

The code uses:
- [RDKit’s ChemicalReaction class](http://rdkit.org/docs/cppapi/classRDKit_1_1ChemicalReaction.html) to reverse a reaction from target to starting materials. Here's an example forward (Pictet-Spengler) and then reverse reaction:
![2023-02-07-Are-the-Starting-Materials-for-Synthesizing-Your-Target-Molecules-Commercially-Available_37_0](https://user-images.githubusercontent.com/1335522/218209676-490349a1-1d5d-4031-a6b0-1cd860daec0a.png)![2023-02-07-Are-the-Starting-Materials-for-Synthesizing-Your-Target-Molecules-Commercially-Available_39_0](https://user-images.githubusercontent.com/1335522/218209668-928298a2-b109-4708-9dc6-e427bcc25e24.png)

- the free [PubChem APIs](https://pubchem.ncbi.nlm.nih.gov/docs/programmatic-access) to determine whether those starting materials are commercially available. PubChem gathers availability from multiple vendors, so we can call just one API source rather than calling dozens of suppliers.
-  [Python’s asyncio library](https://docs.python.org/3/library/asyncio.html) to speed getting API results back by making the API calls concurrently.
- [Semaphore](https://docs.python.org/3/library/asyncio-sync.html#asyncio.Semaphore) to limit the number of concurrent API calls.

**Comment by greglandrum** [comment]
Nice! How do you use the pubchem API to search for commercially available compounds?

**Reply by bertiewooster** [reply]
Thanks! Once you have the PubChem CID (identifier) for the molecule, you use the [PUG VIEW API](https://pubchem.ncbi.nlm.nih.gov/docs/pug-view) to request its Chemical Vendors information. For example, for acetylsalicylic acid (aspirin), the CID is 2244, so the API call to https://pubchem.ncbi.nlm.nih.gov/rest/pug_view/data/compound/2244/XML?heading=Chemical-Vendors returns its vendor information. For compounds _with_ vendor information, the result includes:

```
<TOCHeading>Chemical Vendors</TOCHeading>
<Description>
A list of chemical vendors that sell this compound. ...
</Description>
```

For compounds _without_ vendor information, the result includes:

```
<Code>PUGVIEW.NotFound</Code>
<Message>No data found</Message>
```

which my code keys on to deem a compound not commercially available.

PubChem gathers availability from multiple vendors, so you can call just one API source rather than calling dozens of suppliers'.

**Comment by bertiewooster** [comment]
By the way, is there a way to label the species in a reaction drawing, similar to how you can supply `legends` for MolsToGridImage to label the species below the chemical structure? I would use the reaction drawing here if it were possible to label the species with for example "Reactant available: Yes".
